jobs/fxci-taskcluster-export/fxci_etl/loaders/bigquery.py

```diff
@@ -0,0 +1,66 @@
+from abc import ABC, abstractmethod
+import base64
+from collections import defaultdict
+from dataclasses import asdict, dataclass
+import json
+import os
+from pprint import pprint
+from typing import Any
+
+import dacite
+from google.cloud.bigquery import Client, Table
+
+from fxci_etl.config import Config
+
+
+@dataclass
+class Record(ABC):
+    @classmethod
+    def from_dict(cls, data: dict[str, Any]) -> "Record":
+        return dacite.from_dict(data_class=cls, data=data)
+
+    @classmethod
+    @abstractmethod
+    def table_name(cls) -> str: ...
+
+    @abstractmethod
+    def __str__(self) -> str: ...
+
+
+class BigQueryLoader:
+    def __init__(self, config: Config):
+        self.config = config
+        self._tables = {}
+
+        if config.bigquery.credentials:
+            self.client = Client(
+                json.loads(base64.b64decode(config.bigquery.credentials).decode("utf8"))
+            )
+        else:
+            self.client = Client()
+
+    def get_table(self, name: str) -> Table:
+        if name not in self._tables:
+            bq = self.config.bigquery
+            self._tables[name] = self.client.get_table(
+                f"{bq.project}.{bq.dataset}.{name}"
+            )
+        return self._tables[name]
+
+    def insert(self, records: list[Record] | Record):
+        if isinstance(records, Record):
+            records = [records]
+
+        tables = defaultdict(list)
+        for record in records:
+            tables[record.table_name()].append(record)
+
+        for name, rows in tables.items():
+            table = self.get_table(name)
+            errors = self.client.insert_rows(table, [asdict(row) for row in rows])
+
+            if errors:
+                pprint(errors, indent=2)
```

Comment:
Should this be throwing an exception if there's errors? If not, do we just lose the rows? (The latter seems like it would be problematic if we have large swaths of failures.)

It appears that Pulse messages have already been ack'ed by the time this code runs.

Reply:
Oh yes.. thanks for the call out. I meant to fix that and forgot about it. I need to hook them up to the mechanism that saves the failed events in cloud storage.